modules/mod-deep-learning/DeepLearningEffectBase.cpp

```diff
@@ -0,0 +1,363 @@
+{
+   //Get the samples from the track and put them in the buffer
+   if (!track->GetFloats(buffer, start, len))
+      throw std::runtime_error("An error occurred while copying samples to tensor buffer.");
+
+   // get tensor input from buffer
+   torch::Tensor audio = torch::from_blob(buffer, len,
+                                          torch::TensorOptions().dtype(torch::kFloat));
+   audio = audio.unsqueeze(0); // add channel dimension
+
+   return audio;
+}
+
+torch::Tensor DeepLearningEffectBase::BuildMultichannelTensor(WaveTrack *leader, float *buffer,
+                                                          sampleCount start, size_t len)
+{
+   auto channels = TrackList::Channels(leader);
+   std::vector<torch::Tensor> channelStack;
+
+   // because we're reusing the same buffer, it's important that
+   // we clone the tensor.
+   for (WaveTrack *channel : channels)
+      channelStack.emplace_back(
+          BuildMonoTensor(channel, buffer, start, len).clone());
+
+   return torch::cat(channelStack);
+}
+
+torch::jit::IValue DeepLearningEffectBase::ForwardPassInThread(torch::Tensor input)
+{
+   torch::jit::IValue output;
+
+   std::atomic<bool> done = {false};
+   std::atomic<bool> success = {true};
+
+   // make a copy of the model (in case we need to abort)
+   DeepModelHolder model = this->mModel;
+
+   auto thread = std::thread(
+      [model, &input, &output, &done, &success]()
+      {
+         try
+         {
+            torch::jit::IValue tempOut = model->Forward(input);
+
+            // only write to output tensor if abort was not requested
+            if (success)
+               output = tempOut;
+         }
+         catch (const ModelException &e)
+         {
+            wxLogError(e.what());
+            wxLogDebug(e.what());
+            success = false;
+            output = torch::jit::IValue(torch::zeros_like(input));
+         }
+         done = true;
+      }
+   );
+
+   // wait for the thread to finish
+   while (!done)
+   {
+      if (TrackProgress(mCurrentTrackNum, mCurrentProgress))
+      {
+         // abort if requested
+         success = false;
+
+         // tensor output will be destroyed once the thread is destroyed
+         thread.detach();
+
+         output = torch::jit::IValue(torch::zeros_like(input));
+         return output;
+      }
+
+      ::wxSafeYield();
+      wxMilliSleep(50);
+   }
+   thread.join();
+
+   if (!success)
+   {
+      Effect::MessageBox(XO("An internal error occurred within the neural network model. "
+                        "This model may be broken. Please check the error log for more details"),
+                        wxICON_ERROR);
+   }
+
+   return output;
+}
+
+void DeepLearningEffectBase::TensorToTrack(torch::Tensor waveform, WaveTrack::Holder track,
+                                       double tStart, double tEnd)
+{
+   if (waveform.size(0) != 1)
+      throw Effect::MessageBox(XO("Internal error: input waveform is not mono."));
```

Comment:
This calls a function, which displays a message box and returns int, then throws the int.  That's wrong.

Either do the message box immediately and propagate an error code as the return value, or throw `SimpleMessageBoxException` to delay the display of a message box.